substrate/frame/revive/src/migrations/v1.rs

```diff
@@ -0,0 +1,150 @@
+	/// The storage item that is being migrated from.
+	pub type ContractInfoOf<T: Config> = StorageMap<Pallet<T>, Identity, H160, ContractInfo<T>>;
+}
+
+/// Migrates the items of the [`old::ContractInfoOf`] map into [`crate::AccountInfoOf`].
+pub struct Migration<T: Config>(PhantomData<T>);
+
+impl<T: Config> SteppedMigration for Migration<T> {
+	type Cursor = H160;
+	type Identifier = MigrationId<17>;
+
+	fn id() -> Self::Identifier {
+		MigrationId { pallet_id: *PALLET_MIGRATIONS_ID, version_from: 0, version_to: 1 }
+	}
+
+	fn step(
+		cursor: Option<Self::Cursor>,
+		meter: &mut WeightMeter,
+	) -> Result<Option<Self::Cursor>, SteppedMigrationError> {
+		let required = <T as Config>::WeightInfo::v1_migration_step();
+		if meter.remaining().any_lt(required) {
+			return Err(SteppedMigrationError::InsufficientWeight { required });
+		}
+
+		let mut iter = if let Some(last_key) = cursor {
+			old::ContractInfoOf::<T>::iter_from(old::ContractInfoOf::<T>::hashed_key_for(last_key))
+		} else {
+			old::ContractInfoOf::<T>::iter()
+		}
+		.drain();
+
+		while let Some((last_key, value)) = iter.next() {
+			AccountInfoOf::<T>::insert(
+				last_key,
+				AccountInfo { account_type: value.into(), ..Default::default() },
+			);
+
+			if meter.try_consume(required).is_err() {
+				return Ok(Some(last_key))
+			}
+		}
+
+		Ok(None)
+	}
+
+	#[cfg(feature = "try-runtime")]
+	fn pre_upgrade() -> Result<Vec<u8>, frame_support::sp_runtime::TryRuntimeError> {
+		use codec::Encode;
+
+		// Return the state of the storage before the migration.
+		Ok(old::ContractInfoOf::<T>::iter().collect::<BTreeMap<_, _>>().encode())
+	}
+
+	#[cfg(feature = "try-runtime")]
+	fn post_upgrade(prev: Vec<u8>) -> Result<(), frame_support::sp_runtime::TryRuntimeError> {
+		use codec::Decode;
+
+		// Check the state of the storage after the migration.
+		let prev_map = BTreeMap::<H160, crate::ContractInfo<T>>::decode(&mut &prev[..])
+			.expect("Failed to decode the previous storage state");
+
+		// Check the len of prev and post are the same.
+		assert_eq!(
+			AccountInfoOf::<T>::iter().count(),
+			prev_map.len(),
+			"Migration failed: the number of items in the storage after the migration is not the same as before"
+		);
+
+		for (key, value) in prev_map {
+			let new_value = AccountInfo::<T>::load_contract(&key);
+			assert_eq!(
+				Some(value),
+				new_value,
+				"Migration failed: the value after the migration is not the same as before"
+			);
+		}
+
+		Ok(())
+	}
+}
+
+#[test]
+fn migrate_to_v1() {
+	use crate::{
+		tests::{ExtBuilder, Test},
+		ContractInfo,
+	};
+	ExtBuilder::default().build().execute_with(|| {
+		for i in 0..10u8 {
+			let addr = H160::from([i; 20]);
+			old::ContractInfoOf::<Test>::insert(
+				addr,
+				ContractInfo::new(&addr, 1u32.into(), Default::default()).unwrap(),
+			);
+		}
+
+		let mut cursor = None;
+		let mut weight_meter = WeightMeter::new();
+		while let Some(new_cursor) = Migration::<Test>::step(cursor, &mut weight_meter).unwrap() {
```

Comment:
I think it should be possible to call `pallet_migrations::on_initialize` and `on_finalize` in a loop.

Reply:
yeah don't really care here, just want to test out the step logic itself